Fixed #25790 -- Added option to disable column sort in admin changelist.

ramiro opened the pull request:
https://code.djangoproject.com/ticket/25790

ramiro commented:
Rebased and complete version of #6107.

Wasn't able to force push to sasha0's ticket_25790 branch (permission denied) so that's why I'm opening a new PR.

matthijskooijman commented:
Looks good to me! Are these commits expected to be squashed when merging (would seem appropriate to me)?

ramiro commented:
> Looks good to me! Are these commits expected to be squashed when merging (would seem appropriate to me)?

I hope so. I fixed sasha0 commit when rebasing so it applies cleandy, to try to preserve changes already reviewed in the other PR limited to that commit.

Then I added my changes as new discrete commits to ease their review.

carltongibson commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -1287,6 +1287,22 @@ subclass::
+
+    With ``sortable_by`` you can specify a collection of a subset of these
+    items. You can even disable customization of such ordering by specifying an
+    empty collection.
```
Not 100% against the wording here but I find `...you can specify a collection of a subset of ...` difficult. 
(It doesn't quite read for me.) 

Taking inspiration from `search_fields` (above) maybe something like:

```
With ``sortable_by`` you can specify a subset of these items.
Set ``sortable_by`` to a collection of field names that will be available to sort by.
You can disable customization of such ordering by specifying an
empty collection.
```

ramiro replied:
English isn't my mother tongue and I admit my style can be really baroque :smile: . So feel free to simplify the wording as much as you like. 

One fact we might need to want to make clear is that the items one can use in this settings are not necessarily always model field names, the can also be model or ModelAdmin methods, or functions. Just like in `list_display`.

carltongibson replied:
Yes good point. Perhaps `...collection of column names...` would be enough. (It already mentions the _other callables_ aspect in the preceding paragraph.)

ramiro commented:
> There's now a conflict with a prior change to the release notes.

IMHO this is the only remaining issue. Let me know if you want me to take care of it. IIUC rhese kind of conflict are to be expected and usually the core dev committer is the one who solves it right before the final merge as the release ontes is a rapidly changing document.

carltongibson commented:
Except the tests failed. Gis tests by the looming of it. 

That doesn’t make much sense.

ramiro commented:
buildot, test on xenial.

ramiro commented:
buildbot, test on xenial.

ramiro commented:
I solved the merge conflict, rebased against master and force pushed as an excuse to get Jenkins to re run the tests.

The GIS test failures are still there.

Looking at the failure traceback ( e.g. https://djangoci.com/view/PR%20Builders/job/pull-requests-trusty/16713/database=postgis,label=trusty-pr,python=python3.5/testReport/junit/gis_tests.test_geoip2/GeoIPTest/test05_unicode_response/ ) it looks it's something Tim fixed a couple of weks ago in 66d74676e23c32bc676fb0706af8580b391953b6

It escapes to me why Jenkins is still using the old code for running the PR jobs considering I've rebased the branch against master.

At least another PR is affected by the same problem: e..g https://djangoci.com/view/PR%20Builders/job/pull-requests-trusty/16701/

carltongibson commented:
Hi @ramiro. 

* On the rebase: Yes, would be happy to correct that on merging. (But thanks anyway.) 
* good work on the GIS failure. (Will have to get back to you on that.) 

For me, this PR is Ready for Checkin. Thanks for the effort!

carltongibson commented:
I can't reproduce the failures Jenkins is showing. Indeed, `HEAD` here doesn't contain the problematic code... 

I don't yet know enough about the Jenkins setup to be able to investigate further.  

I've opened #9615 to try a fresh run.

timgraham commented:
Sometimes there are unexplained checkout issues on Jenkins if the pull request contains multiple commits. To retry, you can always amend or squash some commits and force push. I pushed the commits from Carlton's branch here and will review later.

carltongibson commented:
I've rebased this again. I would merge this now.

felixxm commented on django/contrib/admin/options.py:
```diff
@@ -702,6 +707,7 @@ def get_changelist_instance(self, request):
             self.list_max_show_all,
             self.list_editable,
             self,
+            sortable_by
```
Please add trailing comma.

felixxm commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -1296,6 +1296,41 @@ subclass::
+            ssn = models.CharField('SSN', max_length=11)
+
+            def full_name(self):
+                  return self.first_name + ' ' + self.last_name
```
Please fix indentation (remove two extra spaces), i.e.:
```
                return self.first_name + ' ' + self.last_name
```
instead of
```
                  return self.first_name + ' ' + self.last_name
```

felixxm commented on docs/releases/2.1.txt:
```diff
@@ -42,6 +42,10 @@ Minor features
 
 * You can now :ref:`override the the default admin site
   <overriding-default-admin-site>`.
+* New :attr:`.ModelAdmin.sortable_by` option and
```
Please add above a blank line.

felixxm commented on tests/admin_views/admin.py:
```diff
@@ -1069,3 +1070,43 @@ def get_formsets_with_inlines(self, request, obj=None):
+        return ()
+
+
+site6 = admin.AdminSite(name="admin6")
```
Please use single quotes.

felixxm commented on tests/admin_views/tests.py:
```diff
@@ -131,6 +131,7 @@ def setUpTestData(cls):
         cls.chap4 = Chapter.objects.create(title='Chapter 2', content='[ insert contents here ]', book=cls.b2)
         cls.cx1 = ChapterXtra1.objects.create(chap=cls.chap1, xtra='ChapterXtra1 1')
         cls.cx2 = ChapterXtra1.objects.create(chap=cls.chap3, xtra='ChapterXtra1 2')
+        Actor.objects.create(name="Palin", age=27)
```
Please use single quotes.

carltongibson replied:
@felixxm Are you picking these up by eye or linting the diff programatically? If the latter, can I ask your incantation? `flake8` comes out clean with the project config. Ta!

felixxm replied:
@carltongibson By eye, there is no black magic behind 😃 .

carltongibson replied:
@felixxm OK. Thanks. I will just continue to sharpen the nit-picker then 😃

ramiro replied:
Wow. Thank you very much guys. I've learned a few things from this PR discussion. Just pushed fixes for these issues.

Also, where we go we don't need no stinkin' automated linters 😉

felixxm commented on django/contrib/admin/templatetags/admin_list.py:
```diff
@@ -115,13 +116,16 @@ def result_headers(cl):
+            yield {
+                "text": text,
+                "class_attrib": format_html(' class="column-{}"', field_name),
+                "sortable": False,
```
Please use single quotes.
